keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,146 @@
+
+```yaml
+parameters:
+  - name: ports
+    description: Ports
+    type: list
+    default: [ 80, 443 ]
+  - name: topology
+    description: Node topology
+    type: dict
+    default: |
+      - region: us-east-1:
+        replicas: 2
+      - region: us-west-1:
+        replicas: 1
+```
+
+The deployment template then iterates over the items of the respective parameters:
+
+```yaml
+{{ $ports := .Params.ports }}
+{{ range .Params.topology }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  labels:
+    app.kubernetes.io/name: nginx
+spec:
+  replicas: {{ .replicas }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: nginx
+  template:
+    metadata:
+      name: nginx
+      labels:
+        app.kubernetes.io/name: nginx
+    spec:
+      nodeSelector:
+        topology.kubernetes.io/region: {{ .region }}
+      containers:
+      - name: nginx
+        image: nginx:stable
+        ports:
+        {{ range $ports }}
+        - containerPort: {{ . }}
+        {{ end }}
+{{ end }}
+```
+
+### Provide Parameter Validation
+
+Adding types for lists and dictionaries provides access to the items of these structures in templates. Adding types like integers won't provide benefits for templating, but would allow for input value validation. KUDO could, e.g., validate that an integer isn't too large, that a list doesn't have too many entries, or that a string isn't too long. Adding parameter validation is independent from adding parameter types.
+
+Adding validation rules to parameters could look like this:
+
+```yaml
+parameters:
+  - name: replicas
+    description: Number of replicas
+    type: integer
+    minimum: 1
+    maximum: 3
+  - name: instance-name
+    description: Name of the instance
+    regex: ^\w-_*$
+    max-length: 40
+```
+
+### Alternative Considered: Convert YAML in string to dictionaries
+
+A template function `fromYAML` could be used to convert a parameter value to a dictionary. The example above would then change to
+
+```yaml
+{{ $topology := (fromYAML .Params.topology) }}
+{{ range $topology }}
+...
+{{ end }}
+```
+
+In KUDO, this function would work the same way as the `dict` type conversion mentioned above. The difference is that the conversion has to be explicitly done by the operator developer in the template.
+
+### Alternative Considered: Allow Arbitrary Parameter Values
+
+Support for lists or dictionaries could be added by treating every parameter value as YAML. Because a single string is valid YAML, this would be compatible with string parameters. Though, this approach creates problems with Go template pipelines. Some template functions like `eq` no longer work with string parameters that have been converted from YAML, because their resulting type is no longer `string`.
```

Comment:
> string parameters that have been converted from YAML, because their resulting type is no longer `string`.

I'm confused. So are they string parameters or not?

Reply:
Have to double check, I just saw a test using `eq` fail when unmarshalling YAML to `interface{}`.

Reply:
Unmarshalling YAML not only creates `string`, but also `int` and `bool`, and `float`. Currently all our comparisons expect string. This can be resolved by an backwards-incompatible change.

Reply:
Some code to explain this: https://goplay.space/#NaHV_SYYxgC
Of course, this will affect any parameters with YAML type.

I'm testing if a function that converts all types in an `interface{}` to strings might help.